src/template-shadowroot.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+let hasNative: boolean|undefined;
+export function hasNativeDeclarativeShadowRoots(): boolean {
+  if (hasNative === undefined) {
+    const div = document.createElement('div');
+    div.innerHTML = `<div><template shadowroot="open"></template></div>`;
+    hasNative = !!div.firstElementChild!.shadowRoot;
+  }
+  return hasNative;
+}
+
+/*
+ * Traverses the DOM to find all <template> elements with a `shadowroot`
+ * attribute and move their content into a ShadowRoot on their parent element.
+ *
+ * This processing is done bottom up so that when top-level <template>
+ * elements are hydrated, their contents are already hydrated and in the
+ * final correct structure of elements and shadow roots.
+ */
+export const hydrateShadowRoots = (root: ParentNode) => {
+  if (hasNativeDeclarativeShadowRoots()) {
+    return;  // nothing to do
+  }
+
+  // Approaches to try and benchmark:
+  //  - manual walk (current implementation)
+  //  - querySelectorAll
+  //  - TreeWalker
```

Comment:
FYI, TreeWalker doesn't easily let you do operations on exiting a node without additional bookkeeping that's similar to what's here.